webrtc-signal/README.md

```diff
@@ -0,0 +1,218 @@
+   well as one or more proposed configurations for the connection.
+2. The offer is communicated to another peer, called the "answerer".
+3. The answerer generates an "answer" corresponding to the offer. If possible,
+   it selects the best of the connection configurations proposed by the offerer.
+4. The answer is communicated back to the original offerer. If both the offerer
+   and answerer have come to an agreement, a connection is established.
+
+This process may be repeated until both the offerer and the answerer have agreed
+on a configuration for their connection.
+
+Neither WebRTC nor ICE specify _how_ answers and offers should be communicated.
+That detail is left to application developers. The goal of this document is to
+provide a more detailed specification about how answers and offers can be
+communicated to peers using a third-party called a signaler.
+
+## Identity and Authenticity
+
+Each peer that communicates with the signaler is identified by a PeerID.
+Messages sent to the signaler are signed with a private key
+corresponding to that PeerID. This prevents tampering and impersontation. In
+other words, it ensures that a peer can only be connected to the peer that they
+intended to connect to.
+
+## Peer Discovery
+
+This document does not specify how peers should discover one another and is
+designed to be compatible with any peer discovery mechanism. Here are the
+requirements for the WebRTC signaling protocol described in this document:
+
+1. A peer knows the PeerID of the peer it wants to connect to.
+2. Both peers agree to use the same signaler.
+
+It is possible for a signaler itself to implement peer discovery (e.g.,
+using the Rendezvous Protocol), but this is not a strict requirement.
+
+## Multiaddress format
+
+The Signaling Protocol is designed to work over existing libp2p transports. We
+use the following multiaddress format for dialing and listening:
+
+```
+<signaler-multiaddr>/p2p-webrtc-signal/<signaler-peer-id>
+```
+
+Where `<signaler-multiaddr>` is the multiaddress for the signaler (including
+the transport to be used for signaling) and `<signaler-peer-id>` is the
+base58-encoded PeerID of the signaler. `<signaler-multiaddr>` may be omitted to
+use an existing connection to `<signaler-peer-id>` for signaling.
+
+This multiaddress format allows for flexibility in the underlying transport
+used. The signaler can be either centralized (all peers use the same signaler)
+or decentralized (two peers use a common third peer for signaling).
+
+### Examples
+
+- Signaling over WebSockets using an IPv4 address:
+  `/ip4/192.168.1.46/tcp/9000/ws/p2p-webrtc-signal/QmWaWqTtzPCaYnpfxsAAGtrVhNumHqQ7jtdcsFsjvs3csS`
+- Signaling over HTTP using a domain name:
+  `/dns6/signaler.myapp.com/tcp/80/http/p2p-webrtc-signal/QmZbw3TKr3dxhHXiPkbNraWaeGoqPNXAXfAcV8RP2Eqngj`
+- Signaling via a common peer:
+  `/p2p-webrtc-signal/QmWeRHDDiwuGnS4xbjF2zXETucL7xQLjadoaTZ4yJE3hQs`
```

Comment:
Not sure I get this one? Is it missing the ID of the common peer acting as a signaller?

Reply:
It says above

> `<signaler-multiaddr>` may be omitted to use an existing connection to `<signaler-peer-id>` for signaling.

But I'm also not entirely clear on the use case. Does this only work if I've already established a separate connection to the peer I want to reach (e.g. through a circuit relay)?

 Or if we've both connected to any common peer that has the p2p-webrtc-signal capability, could I discover and use that without caring who the signaller peer is?

Reply:
> Or if we've both connected to any common peer that has the p2p-webrtc-signal capability, could I discover and use that without caring who the signaller peer is?

I think that not a good idea, a better solution could be to just indicate on the DHT we are reachable with WebRTC-signal and use a circuit relay to establish the connection.
So for that we need first a ws[s] connection to establish circuit relay and discover the DHT.

Adding some special nodes is a very good way to allow anybody to get a giant control on the network.
When these kind of option are made a very few people activate them, that made people less interested in making these because we know that gonna generate a big traffic.
In fact my solution add a problem about stun server but I think thats more a client side problem (by that I mean not a DHT one) 2, because 2 totaly separated STUN server produce some compatible result (except if you are not in the same network, example: internet and hyperboria), and some STUN server are region locked (China).
We may can just allow peers to run integrated STUN server and broadcast it to the DHT (to be apocalypsical proof) since that give you way less information (a meeting server know who talk to who, can chose who talk to who but can't create fake connection like Qmxxxx... is listening for WebRTC if we require signing these kind of information to a server capable of knowing this ip (so easly matchable with a peer id via DHT), want to connect to a WebRTC node).

**PS**: Meeting server can maybe be used only to broadcast "Qmxxxx is listening for WebRTC, contact him via circuit relay at this node" (signed by Qmxxxx for obvious reason) but in this case why not just putting these directly in DHT listen address of Qmxxxx ?